src/main/scala/com/softwaremill/mqperf/mq/MongoCappedMq.scala

```diff
@@ -0,0 +1,125 @@
+
+  private val db: MongoDatabase = client.getDatabase("mq").withWriteConcern(concern)
+
+  private val queueColl: MongoCollection[Document] = {
+    if (!db.listCollectionNames().iterator().asScala.contains(QueueCollectionName)) {
+      val options = new CreateCollectionOptions()
+        .capped(true)
+        .sizeInBytes(configMap("queue_size_in_bytes").toLong)
+      db.createCollection(QueueCollectionName, options)
+    }
+    db.getCollection(QueueCollectionName)
+  }
+
+  private val cursorStateColl: MongoCollection[Document] = {
+    if (!db.listCollectionNames().iterator().asScala.contains(CursorStateCollectionName)) {
+      db.createCollection(CursorStateCollectionName)
+    }
+    db.getCollection(CursorStateCollectionName)
+  }
+
+  override def createSender(): MqSender = new MongoCappedMqSender(queueColl)
+
+  override def createReceiver(): MqReceiver = new MongoCappedMqReceiver(queueColl, cursorStateColl)
+
+  class MongoCappedMqSender(queueColl: MongoCollection[Document]) extends MqSender {
+
+    /**
+     * Synchronous - must wait for the messages to be sent
+     */
+    override def send(msgs: List[String]): Unit = {
+      val messagesToInsert: java.util.List[Document] = msgs.map { msg =>
+        new Document(MessageField, msg)
+      }.asJava
+      queueColl.insertMany(messagesToInsert)
+    }
+  }
+
+  class MongoCappedMqReceiver(queueColl: MongoCollection[Document], cursorStateColl: MongoCollection[Document]) extends MqReceiver {
+
+    private val nodeId: Int = ObjectId.getGeneratedMachineIdentifier
+
+    private val cursor: MongoCursor[Document] = {
+      val filter = Option(cursorStateColl.find(new BasicDBObject(NodeIdField, nodeId)).first())
+        .map(_.getObjectId(LastDocField))
+        .map(lastDocId => new BasicDBObject(IdField, new BasicDBObject("$gt", lastDocId)))
+        .getOrElse(new BasicDBObject())
+
+      // TODO handle lost capped position
+      queueColl
+        .find(filter)
+        .sort(new BasicDBObject("$natural", 1))
+        .cursorType(CursorType.TailableAwait)
+        .maxAwaitTime(configMap("await_ms").toInt, TimeUnit.MILLISECONDS)
+        .batchSize(configMap("batch_size").toInt)
+        .iterator()
+    }
+
+    override def receive(maxMsgCount: Int): List[(ObjectId, String)] = {
+      doReceive(Nil, maxMsgCount)
+    }
+
+    @tailrec
+    private def doReceive(acc: List[(ObjectId, String)], limit: Int): List[(ObjectId, String)] = {
+      if (limit > 0) {
+        Option(cursor.tryNext()) match {
+          case Some(doc) =>
+            val newElement = doc.getObjectId(IdField) -> doc.getString(MessageField)
+            doReceive(newElement :: acc, limit - 1)
+          case None =>
+            acc
+        }
+      }
+      else {
+        acc
+      }
+    }
+
+    /**
+     * Can be asynchronous
+     */
+    override def ack(ids: List[ObjectId]): Unit = {
+      import scala.concurrent.ExecutionContext.Implicits.global
+      Future {
```

Comment:
Shouldn't we use a separate thread pool here? The `global` may get saturated if we ack fast which may block other activities.